src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewFactory.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Tagging;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Editor.Tagging;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.TextManager.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    /// <summary>
+    /// Responsible for wiring up all the different components of the document outline feature.
+    /// </summary>
+    internal static class DocumentOutlineViewFactory
+    {
+        public static DocumentOutlineView CreateView(
+            ILanguageServiceBroker2 languageServiceBroker,
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListener asyncListener,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            IVsCodeWindow codeWindow)
+        {
+            threadingContext.ThrowIfNotOnUIThread();
+            var (textViewEventSource, textBuffer) = CreateEventSource(asyncListener, editorAdaptersFactoryService, codeWindow);
+            var viewModel = new DocumentOutlineViewModel(languageServiceBroker, asyncListener, textViewEventSource, textBuffer, threadingContext);
+            return new DocumentOutlineView(viewModel, editorAdaptersFactoryService, codeWindow, threadingContext);
+        }
+
+        private static (CompilationAvailableTaggerEventSource, ITextBuffer) CreateEventSource(
+            IAsynchronousOperationListener asyncListener,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            IVsCodeWindow codeWindow)
+        {
+            if (ErrorHandler.Failed(codeWindow.GetLastActiveView(out var textView)))
+            {
+                Debug.Fail("Unable to get the last active text view. IVsCodeWindow implementation we are given is invalid.");
+            }
+
+            var wpfTextView = editorAdaptersFactoryService.GetWpfTextView(textView);
+            Assumes.NotNull(wpfTextView);
+
+            var subjectBuffer = wpfTextView.TextBuffer;
+            var textViewEventSource = new CompilationAvailableTaggerEventSource(
```

Comment:
DocumentOutlint is all syntactic right, and doesn't depend on any semantic information?  If so, get rid of the CompilationAvailableTaggerEventSource as it will just add (significant) overhead.

if we are semantic, we should doc here how we are.

Reply:
Unfortunately, the DocumentSymbol LSP request as its implemented requires semantic information so we need to listen for compilation changes. This can be simplifies if Roslyn's implementation of the DocumentSymbol request is ever changed to just operate off of syntax tree.